src/main/java/org/zalando/intellij/swagger/completion/SwaggerCompletionHelper.java

```diff
@@ -7,7 +7,11 @@
 public class SwaggerCompletionHelper extends CompletionHelper {
 
     private final PathResolver pathResolver;
-
+    
+    public PathResolver getPathResolver() {
```

Comment:
PathResolver for swagger and openapi  do not extend from same parent. They are separate interfaces

Reply:
Ah right. Maybe they should inherit from the same base class, they share quite many methods. This should then be done in the same release, otherwise the change would break the to-be-implemented extensions.

Reply:
Providing new custom fields or custom values for existing fields does not require pathResolver. But in case we need to add custom values for custom fields, path resolver is needed

Here is an example.

```
public class SwaggerCustomValueCompletionFactoryImpl implements SwaggerCustomValueCompletionFactory {
    @Override
    public Optional<ValueCompletion> from(SwaggerCompletionHelper completionHelper, CompletionResultSet completionResultSet) {
        if (completionHelper.completeFormatValue()) {
            return Optional.of(new CustomFormatValueCompletion(completionHelper, completionResultSet));
        } else if (completeCustomRunLevelValue(completionHelper)) {
            return Optional.of(new CustomRunLevelValueCompletion(completionHelper, completionResultSet));
        } else {
            return Optional.empty();
        }
    }

    private static class CustomFormatValueCompletion extends ValueCompletion {

        CustomFormatValueCompletion(final CompletionHelper completionHelper, final CompletionResultSet completionResultSet) {
            super(completionHelper, completionResultSet);
        }

        @Override
        public void fill() {
            CustomSwaggerFieldsValues.format().forEach(this::addValue);
        }
    }

    private static class CustomRunLevelValueCompletion extends ValueCompletion {

        CustomRunLevelValueCompletion(final CompletionHelper completionHelper, final CompletionResultSet completionResultSet) {
            super(completionHelper, completionResultSet);
        }

        @Override
        public void fill() {
            CustomSwaggerFieldsValues.runlevel().forEach(this::addValue);
        }
    }

    private  boolean completeCustomRunLevelValue(CompletionHelper completionHelper) {
        return ((SwaggerCompletionHelper)completionHelper).getPathResolver().hasPath(completionHelper.getPsiElement(), "$.**.x-gw-runlevel");
    }
}
```
And the custom fields and values come from:
```
import java.util.List;

public class CustomSwaggerFieldsValues {

    public static List<Field> root() {
        return ImmutableList.of(

            new StringField("x-gw-combine"),
            new StringField("x-gw-schema-import"),
            new StringField("x-gw-apihandlers"),
            new StringField("x-gw-permissions"),
            new StringField("x-gw-runlevel"),   //custom values for this below
            new StringField("x-gw-serialization")
        );
    }

    public static List<Value> format() {
        return ImmutableList.of(

                new StringValue("x-gw-bigdecimal"),
                new StringValue("x-gw-biginteger"),
                new StringValue("x-gw-money")

        );
    }

    public static List<Value> runlevel() {  //custom values for custom field we added above
        return ImmutableList.of(

                new StringValue("starting"),
                new StringValue("maintenance"),
                new StringValue("daemons"),
                new StringValue("multiuser")

        );
    }
}

```

And we register the extensions in plugin.xml ofcourse

Reply:
Also, I updated a comment below about the tests.

Reply:
Thanks for the clarification. In this case, I think we can just add a method like this to the `CompletionHelper`

```java
public boolean hasPath(final String pathExpression) {
    return new PathFinder().isInsidePath(psiElement, pathExpression);
}
```

Then we wouldn't need the `getPathResolver()` and `getPsiElement()` methods. What do you think?

Reply:
Sounds good. 
I will do it tonight 
Also, would you like me to do these extension points for openapi as well? I guess it will be pretty much the same

Reply:
Ok great. Looking forward to releasing this :)

Reply:
Hi there
I looked into the code for JSON completion. 
1. Not sure if the current code completion works for *.json files. It is the default completion for true, false and null values only. The completion contributor checks for swagger files. May be I am missing something here. Could you give an example of Json code completion? Thanks
2. Json completionContributor extension point already exists in IntelliJ out of the box as Json is a bundled plugin. A little while ago, I did implement the extension point in our product. Do you still want to have an extension point in swagger plugin when it is there in the bundled json?

Please let me know what you think. May be I am missing something.

Reply:
About the point #2 above.. let me correct
The extension point for Json does not exist specifically. But since Intellij provides the Psi stuff for Json, a user can extend the same CompletionContributor for Json files. 
For yaml files, the Psi structure does not exist in Intellij and that is where your plugin does all the work.
Please let me know if I am not very clear. We can have a chat over email too. Thanks